shrinklvm.py

```diff
@@ -0,0 +1,1012 @@
+    assert(len(mda_data) <= lvm.mda[0].size)
+    mda_header += struct.pack('<QQII', SECTOR_SIZE, len(mda_data), crc, 0)
+    mda_header += '\x00' * (SECTOR_SIZE - 4 - len(mda_header))
+    crc = calc_crc(mda_header)
+    sector = struct.pack('<I', crc) + mda_header
+
+    # Write metadata header sector and text config
+    lvm.file.seek(offset + lvm.mda[0].offset, os.SEEK_SET)
+    lvm.file.write(sector)
+    lvm.file.write(mda_data)
+
+
+def shrink_partition(drive, partno, start_shrunk_sectors, end_shrunk_sectors):
+    '''Shrink the partition to match the new LVM size'''
+
+    tool = PartitionTool(drive)
+    old_partition = tool.partitions[partno]
+    part_start = old_partition['start']
+    part_size = old_partition['size']
+
+    logger.log('Old partition: {} {}'.format(part_start, part_size))
+    part_start += start_shrunk_sectors
+    part_size -= start_shrunk_sectors
+    part_size -= end_shrunk_sectors
+    logger.log('New partition: {} {}'.format(part_start, part_size))
+
+    tool.deletePartition(partno)
+    tool.createPartition(old_partition['id'], sizeBytes=(part_size * tool.sectorSize),
+                         number=partno, startBytes=(part_start * tool.sectorSize),
+                         active=old_partition['active'])
+
+    tool.commit(log=True)
+
+
+def shrink_lvm(drive, partno, start_space, end_space, progress_callback):
+    '''Shrink an LVM partition
+
+    This operation moves the start of the LVM volume forward by start_space
+    bytes, moves the end backward by end_space bytes, and then resizes the
+    partition to fit. The start is moved forward precisely to ensure that
+    alignment is respected. This is done by increasing the size of the metadata
+    area. The end_space is rounded up to the nearest extent.
+    '''
+
+    logger.log('Reading LVM header...')
+    lvm = read_header(partitionDevice(drive, partno))
+    logger.log('Original LVM config: {}'.format(lvm.text_config))
+
+    logger.log('Parsing LVM config...')
+    parse_config(lvm)
+
+    logger.log('Checking LVM device...')
+    check_lvm(lvm)
+
+    # Create a sorted allocation table of the segments
+    alloc_table = lvm_to_alloc_table(lvm)
+    logger.log('Initial allocation table: {}'.format(alloc_table))
+
+    extent_size = int(lvm.vgs[0].get_config('extent_size')) * SECTOR_SIZE
+    pe_count = int(lvm.vgs[0].pvs[0].get_config('pe_count'))
+    start_shrunk_extents = int(math.ceil(start_space / float(extent_size)))
+    end_shrunk_extents = int(math.ceil(end_space / float(extent_size)))
+    regions = [ReservedRegion(0, start_shrunk_extents)]
+    if end_shrunk_extents > 0:
+        regions.append(ReservedRegion(pe_count - end_shrunk_extents, end_shrunk_extents))
+
+    # If we're shrinking the start by an amount that is not a multiple of the extent size
+    # increase the size of the metadata area to compensate
+    extra_metadata_sectors = (start_shrunk_extents * extent_size - start_space) / SECTOR_SIZE
```

Comment:
Just to be safe I would use `// SECTOR_SIZE`. I suppose for some weird reason LVM2 assume 512 sector size that is that is use that assumption as base unit instead of assuming the disk has always that sector size.

Reply:
I'll fix that.

LVM2 (at least with the version we use) does seem to have a hardcoded sector size of 512: https://github.com/lvmteam/lvm2/blob/bb384f84880a183ee58a18935851eece25cdada4/lib/metadata/metadata-exported.h#L32